File: Problems/Fruit Into Baskets.py
# Use sliding windows to get range of fruits in basket
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        basket = {}
        max_fruit = 0
        start = 0

        for end in range(len(fruits)):
            basket[fruits[end]] = basket.get(fruits[end], 0) + 1
            while len(basket) > 2:
                basket[fruits[start]] -= 1
                if basket[fruits[start]] == 0:
                    del basket[fruits[start]]
                start += 1
            
            max_fruit = max(max_fruit, end - start + 1)
        
        return max_fruit

File: Problems/Fruit Into Baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        Map<Integer, Integer> state = new HashMap<>();
        int res = 0;

        int l = 0;
        int r = 0;
        while (r < fruits.length) {
            if (state.size() < 2 || state.containsKey(fruits[r])) {
                state.merge(fruits[r], 1, Integer::sum);
                ++r;
                res = Math.max(res, r-l);
            } else {
                int decremented = state.get(fruits[l]) - 1;
                if (decremented == 0) {
                    state.remove(fruits[l]);
                } else {
                    state.put(fruits[l], decremented);
                }
                ++l;
            }   
        }

        return res;
    }
}
